File: frontend/src/types/timeline.ts
export type EntryType = 'detection' | 'analysis' | 'containment' | 'evidence' | 'comms' | 'note'

export interface Attachment {
  id: string
  original_name: string
  mime_type: string | null
  file_size: number | null
  sha256: string
}

export interface TimelineEntry {
  id: string
  incident_id: string
  author_id: string
  author_name?: string
  entry_type: EntryType
  occurred_at: string
  description: string
  source: string | null
  is_pinned: boolean
  metadata: Record<string, unknown>
  created_at: string
  attachments?: Attachment[]
}

export interface CreateTimelineEntryPayload {
  entry_type: EntryType
  occurred_at: string
  description: string
  source?: string
}

export interface UpdateTimelineEntryPayload {
  entry_type?: EntryType
  occurred_at?: string
  description?: string
  source?: string
}

export const ENTRY_TYPE_CONFIG: Record<EntryType, {
  label: string
  dotClass: string
  badgeClass: string
  icon: string
}> = {
  detection: {
    label: 'Detection',
    dotClass: 'dot-detection',
    badgeClass: 'badge-detection',
    icon: '🔍',
  },
  analysis: {
    label: 'Analysis',
    dotClass: 'dot-analysis',
    badgeClass: 'badge-analysis',
    icon: '🔬',
  },
  containment: {
    label: 'Containment',
    dotClass: 'dot-containment',
    badgeClass: 'badge-containment',
    icon: '🛡️',
  },
  evidence: {
    label: 'Evidence',
    dotClass: 'dot-evidence',
    badgeClass: 'badge-evidence',
    icon: '📎',
  },
  comms: {
    label: 'Comms',
    dotClass: 'dot-comms',
    badgeClass: 'badge-comms',
    icon: '💬',
  },
  note: {
    label: 'Note',
    dotClass: 'dot-note',
    badgeClass: 'badge-note',
    icon: '📝',
  },
}

File: backend/app/schemas/timeline.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel


class TimelineEntryCreate(BaseModel):
    entry_type: Literal["detection", "analysis", "containment", "evidence", "comms", "note"]
    occurred_at: datetime
    description: str
    source: str = "manual"
    is_pinned: bool = False
    metadata: dict = {}


class TimelineEntryUpdate(BaseModel):
    entry_type: Literal["detection", "analysis", "containment", "evidence", "comms", "note"] | None = None
    occurred_at: datetime | None = None
    description: str | None = None
    source: str | None = None
    is_pinned: bool | None = None
    metadata: dict | None = None


class AttachmentBrief(BaseModel):
    id: UUID
    original_name: str
    mime_type: str | None
    file_size: int | None
    sha256: str

    model_config = {"from_attributes": True}


class TimelineEntryOut(BaseModel):
    id: UUID
    incident_id: UUID
    author_id: UUID | None
    entry_type: str
    occurred_at: datetime
    description: str
    source: str
    is_pinned: bool
    metadata: dict
    created_at: datetime
    updated_at: datetime
    attachments: list[AttachmentBrief] = []

    model_config = {"from_attributes": True}

    @classmethod
    def model_validate(cls, obj, *args, **kwargs):
        if hasattr(obj, "metadata_"):
            obj.__dict__["metadata"] = obj.metadata_
        return super().model_validate(obj, *args, **kwargs)
